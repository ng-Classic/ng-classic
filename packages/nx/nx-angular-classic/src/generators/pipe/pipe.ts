import type { Tree } from '@nx/devkit';
import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  names,
} from '@nx/devkit';
import { addToNgModule, findModule } from '../utils';
import { normalizeOptions, validateOptions } from './lib';
import type { Schema } from './schema';

export async function pipeGenerator(tree: Tree, rawOptions: Schema) {
  validateOptions(tree, rawOptions);
  const options = normalizeOptions(tree, rawOptions);

  const pipeNames = names(options.name);

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    options.directory,
    {
      pipeClassName: pipeNames.className,
      pipeFileName: pipeNames.fileName,
      pipePropertyName: pipeNames.propertyName,
      standalone: options.standalone,
      tpl: '',
    }
  );

  if (options.skipTests) {
    const pathToSpecFile = joinPathFragments(
      options.directory,
      `${pipeNames.fileName}.pipe.spec.ts`
    );

    tree.delete(pathToSpecFile);
  }

  if (!options.skipImport && !options.standalone) {
    const modulePath = findModule(tree, options.path, options.module);
    addToNgModule(
      tree,
      options.path,
      modulePath,
      pipeNames.fileName,
      `${pipeNames.className}Pipe`,
      `${pipeNames.fileName}.pipe`,
      'declarations',
      options.flat,
      options.export
    );
  }

  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default pipeGenerator;
