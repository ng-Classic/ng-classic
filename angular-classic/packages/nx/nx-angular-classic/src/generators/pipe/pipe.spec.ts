import { addProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { pipeGenerator } from './pipe';
import type { Schema } from './schema';

describe('pipe generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();

    addProjectConfiguration(tree, 'test', {
      root: 'test',
      sourceRoot: 'test/src',
      projectType: 'application',
    });

    tree.write(
      'test/src/app/test.module.ts',
      `import {NgModule} from "@angular-classic/core";
    @NgModule({
      imports: [],
      declarations: [],
      exports: []
    })
    export class TestModule {}`
    );
  });

  it('should generate a pipe with test files and attach to the NgModule automatically', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree);

    // ASSERT
    expect(tree.read('test/src/app/test.pipe.ts', 'utf-8')).toMatchSnapshot();
    expect(
      tree.read('test/src/app/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should import the pipe correctly when flat=false', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, { flat: false });

    // ASSERT
    expect(
      tree.read('test/src/app/test/test.pipe.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/src/app/test/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should not import the pipe when standalone=true', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, { standalone: true });

    // ASSERT
    expect(tree.read('test/src/app/test.pipe.ts', 'utf-8')).toMatchSnapshot();
    expect(
      tree.read('test/src/app/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should import the pipe correctly when flat=false and path is nested deeper', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, {
      flat: false,
      path: 'test/src/app/my-pipes',
    });

    // ASSERT
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should export the pipe correctly when flat=false and path is nested deeper', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, {
      flat: false,
      path: 'test/src/app/my-pipes',
      export: true,
    });

    // ASSERT
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should not import the pipe when skipImport=true', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, {
      flat: false,
      path: 'test/src/app/my-pipes',
      skipImport: true,
    });

    // ASSERT
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should not generate test file when skipTests=true', async () => {
    // ARRANGE

    // ACT
    await generatePipeWithDefaultOptions(tree, {
      flat: false,
      path: 'test/src/app/my-pipes',
      skipTests: true,
    });

    // ASSERT
    expect(
      tree.read('test/src/app/my-pipes/test/test.pipe.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.exists('test/src/app/my-pipes/test/test.pipe.spec.ts')
    ).toBeFalsy();
    expect(tree.read('test/src/app/test.module.ts', 'utf-8')).toMatchSnapshot();
  });
});

async function generatePipeWithDefaultOptions(
  tree: Tree,
  overrides: Partial<Schema> = {}
) {
  await pipeGenerator(tree, {
    name: 'test',
    project: 'test',
    flat: true,
    ...overrides,
  });
}
