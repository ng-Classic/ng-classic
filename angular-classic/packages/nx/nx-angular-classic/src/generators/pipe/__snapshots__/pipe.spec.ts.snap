// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipe generator should export the pipe correctly when flat=false and path is nested deeper 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should export the pipe correctly when flat=false and path is nested deeper 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should export the pipe correctly when flat=false and path is nested deeper 3`] = `
"import { NgModule } from '@angular-classic/core';
import { TestPipe } from './my-pipes/test/test.pipe';
@NgModule({
  imports: [],
  declarations: [TestPipe],
  exports: [TestPipe],
})
export class TestModule {}
"
`;

exports[`pipe generator should generate a pipe with test files and attach to the NgModule automatically 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should generate a pipe with test files and attach to the NgModule automatically 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should generate a pipe with test files and attach to the NgModule automatically 3`] = `
"import { NgModule } from '@angular-classic/core';
import { TestPipe } from './test.pipe';
@NgModule({
  imports: [],
  declarations: [TestPipe],
  exports: [],
})
export class TestModule {}
"
`;

exports[`pipe generator should import the pipe correctly when flat=false 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should import the pipe correctly when flat=false 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should import the pipe correctly when flat=false 3`] = `
"import { NgModule } from '@angular-classic/core';
import { TestPipe } from './test/test.pipe';
@NgModule({
  imports: [],
  declarations: [TestPipe],
  exports: [],
})
export class TestModule {}
"
`;

exports[`pipe generator should import the pipe correctly when flat=false and path is nested deeper 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should import the pipe correctly when flat=false and path is nested deeper 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should import the pipe correctly when flat=false and path is nested deeper 3`] = `
"import { NgModule } from '@angular-classic/core';
import { TestPipe } from './my-pipes/test/test.pipe';
@NgModule({
  imports: [],
  declarations: [TestPipe],
  exports: [],
})
export class TestModule {}
"
`;

exports[`pipe generator should not generate test file when skipTests=true 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should not generate test file when skipTests=true 2`] = `
"import { NgModule } from '@angular-classic/core';
import { TestPipe } from './my-pipes/test/test.pipe';
@NgModule({
  imports: [],
  declarations: [TestPipe],
  exports: [],
})
export class TestModule {}
"
`;

exports[`pipe generator should not import the pipe when skipImport=true 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should not import the pipe when skipImport=true 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should not import the pipe when skipImport=true 3`] = `
"import { NgModule } from '@angular-classic/core';
@NgModule({
  imports: [],
  declarations: [],
  exports: [],
})
export class TestModule {}
"
`;

exports[`pipe generator should not import the pipe when standalone=true 1`] = `
"import { Pipe, PipeTransform } from '@angular-classic/core';

@Pipe({
  name: 'test',
  standalone: true,
})
export class TestPipe implements PipeTransform {
  transform(value: unknown, ...args: unknown[]): unknown {
    return null;
  }
}
"
`;

exports[`pipe generator should not import the pipe when standalone=true 2`] = `
"import { TestPipe } from './test.pipe';

describe('TestPipe', () => {
  it('create an instance', () => {
    const pipe = new TestPipe();
    expect(pipe).toBeTruthy();
  });
});
"
`;

exports[`pipe generator should not import the pipe when standalone=true 3`] = `
"import { NgModule } from '@angular-classic/core';
@NgModule({
  imports: [],
  declarations: [],
  exports: [],
})
export class TestModule {}
"
`;
